WordPress/gutenberg
# State: Refactor post dirty state as higher-order reducer

Fixes #3296 
Related: #1996

This pull request seeks to refactor the state behavior for determining whether a post has unsaved changes. In some ways it reverts to characteristics prior to #1996 in that the proposed changes here no longer perform a diff between the current and saved variations of a post. This is proposed for two reasons:

- As of #3178, we can no longer assume to have history reset at each save, so with the implementation of the [`isEditedPostDirty` selector](https://github.com/WordPress/gutenberg/blob/35fcb3244afc1363cdb303158309f32a061ee95f/editor/selectors.js#L177-L226) we'd otherwise need a reference to blocks at the time of the last save (either tracking history at specific action intervals, or a separate reducer)
- The `isEditedPostDirty` selector, while memoized, contains non-trivial logic (multiple deep object comparison), and is cachebust on a regular basis (each edit), thus potentially harming performance.

As implemented here, the basic idea is that we check references on `state.editor` before and after any action is dispatched to the store, _using a `withChangeDetection` Redux higher-order reducer. Since this reducer value encompasses any edits which occur, a strict reference equality check should be sufficient for determining if an edit has been made. This is made more challenging by:

- Known editor resets (at save, at initial load)
- Cannot reflect Saved state when manually undo-ing edits (like demonstrated in #1996)

The latter of these, while "nice to have", is not as critical a feature as I might have originally assumed.

__Implementation notes:__

~The state structuring here partially reflects the proposed Ducks pattern of #3012, originally planned to be iterated toward as part of #3030, but prioritized into these changes as a critical bug fix.~

__Testing instructions:__

Repeat testing instructions from #3296, verifying Expected Result.

Verify that Save Draft button only appears at expected intervals:

- Not on initial page load
- When changing block order
- When changing block attributes / content
- When inserting blocks
- When changing post attributes (title, date, etc.)
- When undoing or redoing
